internal/testutils/testserver.go

```diff
@@ -51,59 +52,108 @@ func (t TestServer) FaultyClient() *http.Client {
 		}
 
-		if len(responses) <= testServer.calls-1 {
-			t.Fail()
-			t.Fatalf("Exceeded number of calls to test server (expected: %d)", len(responses))
+		responseDef := responses[testServer.calls-1]
+		handlers := map[string]func(*testing.T, *http.Request) Response{
+			http.MethodGet:    responseDef.Get,
+			http.MethodPost:   responseDef.Post,
+			http.MethodPut:    responseDef.Put,
+			http.MethodDelete: responseDef.Delete,
 		}
 
-		if res, found := responses[testServer.calls-1][req.Method]; found {
-			if res.ValidateRequestFunc != nil {
-				res.ValidateRequestFunc(req)
-			}
-			rw.WriteHeader(res.ResponseCode)
-			_, _ = rw.Write([]byte(res.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
-		} else {
-			t.Errorf("unexpected HTTP method call: %s", req.Method)
+		handlerFunc, found := handlers[req.Method]
+		if !found {
+			panic(fmt.Sprintf("No %s method defined for server call nr. %d", req.Method, testServer.calls))
 		}
+		response := handlerFunc(t, req)
+		rw.WriteHeader(response.ResponseCode)
+		_, _ = rw.Write([]byte(response.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
+		responseDef.Validate(t, req)
 	}
 	testServer.Server = httptest.NewServer(http.HandlerFunc(handler))
 	return testServer
 }
 
+type Response struct {
+	ResponseCode int
+	ResponseBody string
+}
+
+type ResponseDef struct {
+	GET             func(*testing.T, *http.Request) Response
+	PUT             func(*testing.T, *http.Request) Response
+	POST            func(*testing.T, *http.Request) Response
+	DELETE          func(*testing.T, *http.Request) Response
+	ValidateRequest func(*testing.T, *http.Request)
+}
+
+func (r ResponseDef) Get(t *testing.T, req *http.Request) Response {
+	if r.GET == nil {
+		panic("GET() function not defined")
+	}
+	return r.GET(t, req)
+}
+func (r ResponseDef) Put(t *testing.T, req *http.Request) Response {
+	if r.PUT == nil {
+		panic("PUT() function not defined")
+	}
+	return r.PUT(t, req)
+}
+func (r ResponseDef) Post(t *testing.T, req *http.Request) Response {
+	if r.POST == nil {
+		panic("POST() function not defined")
+	}
+	return r.POST(t, req)
+}
+func (r ResponseDef) Delete(t *testing.T, req *http.Request) Response {
+	if r.DELETE == nil {
+		panic("DELETE() function not defined")
+	}
+	return r.DELETE(t, req)
+}
+
+func (r ResponseDef) Validate(t *testing.T, req *http.Request) {
+	if r.ValidateRequest != nil {
+		r.ValidateRequest(t, req)
+	}
+}
+
 // ErrorTransport is custom transport that always produces a simulated network error.
 type ErrorTransport struct{}
 
 // RoundTrip implements the RoundTripper interface and returns a simulated network error.
 func (t *ErrorTransport) RoundTrip(_ *http.Request) (*http.Response, error) {
 	return nil, errors.New("simulated network error")
 }
+
+func checkExactlyOneHandlerSet(def ResponseDef) bool {
```

Comment:
nit: technically the name is incorrect. It allows zero handlers as well - not sure if it even should though? 
If I want to assert that no API calls happen, a testserver with zero responses would also make the test fail if it gets one, right?